server/src/main/java/io/crate/auth/HttpAuthUpstreamHandler.java

```diff
@@ -95,7 +99,26 @@ protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Except
     private void handleHttpRequest(ChannelHandlerContext ctx, HttpRequest request) {
         SSLSession session = getSession(ctx.channel());
         Credentials credentials = credentialsFromRequest(request, session, settings);
-        String username = credentials.username();
+
+        String username;
+        var decodedJwt = credentials.decodedToken();
+        if (decodedJwt != null) {
+            Role role = roles.findUser(
+                decodedJwt.getIssuer(),
+                decodedJwt.getClaim("username").asString()
+            );
+            if (role != null) {
+                username = role.name();
+            } else {
+                // Set to empty string to avoid NPE
+                // and also align with behavior of Header.extractCredentialsFromHttpAuthHeader where we set an empty user as fallback.
+                username = "";
+            }
+            credentials.setUsername(username);
+        } else {
+            username = credentials.username();
+        }
```

Comment:
Thx, added an NPE check in https://github.com/crate/crate/pull/15551/commits/e5d26ba527de5838e561ff460a7f4ae59b717110

We don't need such checks in `ClientCertAuth`, `PasswordAuthenticationMethod` and in `SqlHttpHandler` because we either explicitly fallback to `AuthSettings.AUTH_TRUST_HTTP_DEFAULT_HEADER` or fetch username from Basic header. 

For empty or null header we set empty credentials. And 
`authService.resolveAuthenticationType(username, connectionProperties) `is also null safe for both implementaitons.

Reply:
We still set an empty string as the `username` instead of `null` at https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/protocols/http/Headers.java#L70. 
Can we change this please and also remove any `NotNull` annotations at `Credentials` for the username?
I don't see any reason why we need to differentiate between a `null` or empty username.

Reply:
Thx, did that and use null as default value for password (we always do null check on close and there is no reason to create dummy empty char array.

BTW. `setUsername` still has `NotNull` since it's really not null on both usages (we take it from a Role after doing null check)